# FİNAL PROJE                                                                         ##### Şevval Şereflican

### 1.A.REGRESYON PROBLEMİ

In [4]:
# Gerekli kütüphaneler
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [6]:
# Veriyi yükleyelim
df = pd.read_csv('train.csv')  

# İlk 5 satıra bakalım
df.head()


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [8]:
# Veri seti hakkında genel bilgi
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [10]:
# Eksik veri kontrolü
df.isnull().sum().sort_values(ascending=False).head(20)

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageYrBlt       81
GarageCond        81
GarageType        81
GarageFinish      81
GarageQual        81
BsmtFinType2      38
BsmtExposure      38
BsmtQual          37
BsmtCond          37
BsmtFinType1      37
MasVnrArea         8
Electrical         1
Id                 0
dtype: int64

In [12]:
# Çok fazla eksik olan sütunları çıkaralım
df = df.drop(['PoolQC', 'MiscFeature', 'Alley', 'Fence'], axis=1)

# 'Id' sütununu çıkaralım
df = df.drop('Id', axis=1)

# Kategorik değişkenlerde eksikleri 'None' ile doldur
cat_cols = df.select_dtypes(include=['object']).columns
for col in cat_cols:
    df[col] = df[col].fillna('None')

# Sayısal değişkenlerde eksikleri medyan ile doldur
num_cols = df.select_dtypes(include=[np.number]).columns
for col in num_cols:
    df[col] = df[col].fillna(df[col].median())

# Eksik verilerin tekrar kontrolü
print(f"Toplam eksik veri sayısı: {df.isnull().sum().sum()}")


Toplam eksik veri sayısı: 0


In [14]:
# Kategorik değişkenleri one-hot encoding ile sayısala çevirelim
df_encoded = pd.get_dummies(df, drop_first=True)

print(f"Orijinal veri seti sütun sayısı: {df.shape[1]}")
print(f"One-hot encoding sonrası sütun sayısı: {df_encoded.shape[1]}")

# Satış fiyatı hedef değişkeni olarak ayıralım
X = df_encoded.drop('SalePrice', axis=1)
y = df_encoded['SalePrice']

# Veri setini eğitim ve test olarak ayıralım
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)


Orijinal veri seti sütun sayısı: 76
One-hot encoding sonrası sütun sayısı: 248


In [16]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Model tanımlama
lr_model = LinearRegression()

# Modeli eğitme
lr_model.fit(X_train, y_train)

# Test setinde tahmin
y_pred = lr_model.predict(X_test)

# Performans ölçümleri
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Lineer Regresyon Performansı:")
print(f"MSE: {mse:.2f}")
print(f"MAE: {mae:.2f}")
print(f"R2: {r2:.4f}")


Lineer Regresyon Performansı:
MSE: 2757705912.80
MAE: 20656.42
R2: 0.6405


### R² değeri yaklaşık %64, bu da modelin satış fiyatındaki değişimin %64’ünü açıklayabildiği anlamına gelir. MSE ve MAE ise fiyatların birimlerine göre yüksek görünüyor, ama bu normal çünkü konut fiyatları genelde büyük değerlerdir.


In [20]:
from sklearn.neighbors import KNeighborsRegressor

# Model tanımlama (k=5 varsayılan)
knn_model = KNeighborsRegressor()

# Modeli eğitme
knn_model.fit(X_train, y_train)

# Tahmin
y_pred_knn = knn_model.predict(X_test)

# Performans ölçümü
mse_knn = mean_squared_error(y_test, y_pred_knn)
mae_knn = mean_absolute_error(y_test, y_pred_knn)
r2_knn = r2_score(y_test, y_pred_knn)

print("KNN Regressor Performansı:")
print(f"MSE: {mse_knn:.2f}")
print(f"MAE: {mae_knn:.2f}")
print(f"R2: {r2_knn:.4f}")


KNN Regressor Performansı:
MSE: 2273211595.89
MAE: 28141.80
R2: 0.7036


### KNN R² performansı Lineer Regresyon’dan daha iyi (%70 civarı), ama MAE biraz daha yüksek olmuş.

In [23]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, VotingRegressor
from sklearn.svm import SVR

# 1) Decision Tree
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)
mse_dt = mean_squared_error(y_test, y_pred_dt)
mae_dt = mean_absolute_error(y_test, y_pred_dt)
r2_dt = r2_score(y_test, y_pred_dt)

# 2) Random Forest
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
mse_rf = mean_squared_error(y_test, y_pred_rf)
mae_rf = mean_absolute_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

# 3) SVR (varsayılan parametrelerle, biraz yavaş olabilir)
svr_model = SVR()
svr_model.fit(X_train, y_train)
y_pred_svr = svr_model.predict(X_test)
mse_svr = mean_squared_error(y_test, y_pred_svr)
mae_svr = mean_absolute_error(y_test, y_pred_svr)
r2_svr = r2_score(y_test, y_pred_svr)

# 4) Voting Regressor (LR, RF ve SVR modellerinin birleşimi)
voting_model = VotingRegressor(
    estimators=[('lr', lr_model), ('rf', rf_model), ('svr', svr_model)]
)
voting_model.fit(X_train, y_train)
y_pred_voting = voting_model.predict(X_test)
mse_voting = mean_squared_error(y_test, y_pred_voting)
mae_voting = mean_absolute_error(y_test, y_pred_voting)
r2_voting = r2_score(y_test, y_pred_voting)

# Sonuçları yazdır
print("Decision Tree Performansı:")
print(f"MSE: {mse_dt:.2f}, MAE: {mae_dt:.2f}, R2: {r2_dt:.4f}\n")

print("Random Forest Performansı:")
print(f"MSE: {mse_rf:.2f}, MAE: {mae_rf:.2f}, R2: {r2_rf:.4f}\n")

print("SVR Performansı:")
print(f"MSE: {mse_svr:.2f}, MAE: {mae_svr:.2f}, R2: {r2_svr:.4f}\n")

print("Voting Regressor Performansı:")
print(f"MSE: {mse_voting:.2f}, MAE: {mae_voting:.2f}, R2: {r2_voting:.4f}")


Decision Tree Performansı:
MSE: 1841929467.72, MAE: 27535.11, R2: 0.7599

Random Forest Performansı:
MSE: 847842377.62, MAE: 17758.45, R2: 0.8895

SVR Performansı:
MSE: 7859359444.93, MAE: 59556.73, R2: -0.0246

Voting Regressor Performansı:
MSE: 2031654028.06, MAE: 25835.63, R2: 0.7351


##### Decision Tree: R² : %76, fena değil.

##### Random Forest: R² : %89, şu an en iyi model bu görünüyor, hem MSE hem MAE düşük.

##### SVR: Çok kötü performans (-0.02 R²) , parametre ayarı kesinlikle gerekli.(En basit düzeyde de parametre kullansamda kod çok yavaş olduğundan
##### hep çalışmakta kaldı dolayısıyla gereksede parametre uygulamadım bu nedenle de  Svr büyük veri setlerinde çok verimli bir model değildir.)

##### Voting Regressor (LR + RF + SVR): R² :%73, SVR’nin kötü etkisi olabilir, optimize etmek gerek.

In [29]:
from sklearn.model_selection import GridSearchCV

param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2]
}

rf = RandomForestRegressor(random_state=42)
grid_search_rf = GridSearchCV(estimator=rf, param_grid=param_grid_rf, cv=3, n_jobs=-1, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)

print("En iyi parametreler (Random Forest):", grid_search_rf.best_params_)

best_rf = grid_search_rf.best_estimator_

# Test performansı
y_pred_best_rf = best_rf.predict(X_test)
print("Optimize edilmiş Random Forest Performansı:")
print(f"MSE: {mean_squared_error(y_test, y_pred_best_rf):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_best_rf):.2f}")
print(f"R2: {r2_score(y_test, y_pred_best_rf):.4f}")


En iyi parametreler (Random Forest): {'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 5, 'n_estimators': 100}
Optimize edilmiş Random Forest Performansı:
MSE: 845773334.85
MAE: 17739.75
R2: 0.8897


##### Parametre optimizasyonu sonrası MSE ve MAE biraz düştü, bu da tahminlerin gerçek değerlere biraz daha yakınlaştığını gösterir.

##### R² skoru da hafifçe yükseldi, yani model verinin varyansını biraz daha iyi açıklıyor.

##### Ancak, performans iyileşmesi çok büyük değil, yani model zaten varsayılan haliyle de oldukça iyiydi.

##### Fakat buu küçük iyileşme bile özellikle büyük veri setlerinde veya kritik uygulamalarda anlamlı olabilir.

##### Optimize edilen parametreler genellikle modeli biraz daha genelleştirip aşırı öğrenmeyi engellemeye yarar.



In [ ]:
# Kodun çıktılarına ulaşamasamda kodu koymak istedim.
from sklearn.svm import SVR
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from scipy.stats import uniform

# Eğitim verisinin küçük bir alt kümesini kullan (performans için)
X_sample = X_train[:1000]
y_sample = y_train[:1000]

# Daha dar ama etkili parametre aralığı
param_dist_svr = {
    'kernel': ['linear', 'rbf'],
    'C': uniform(1, 50),
    'gamma': ['scale', 'auto']
}

svr = SVR()

random_search_svr = RandomizedSearchCV(
    estimator=svr,
    param_distributions=param_dist_svr,
    n_iter=10,               # 10 kombinasyon
    cv=3,
    n_jobs=-1,
    scoring='neg_mean_squared_error',
    random_state=42,
    verbose=1
)

# Fit (hızlı örneklemle)
random_search_svr.fit(X_sample, y_sample)

# En iyi modelle tam test setinde tahmin
print("En iyi parametreler (SVR):", random_search_svr.best_params_)

best_svr = random_search_svr.best_estimator_
y_pred_best_svr = best_svr.predict(X_test)

print("Optimize edilmiş SVR Performansı:")
print(f"MSE: {mean_squared_error(y_test, y_pred_best_svr):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_best_svr):.2f}")
print(f"R2: {r2_score(y_test, y_pred_best_svr):.4f}")

In [54]:
from sklearn.ensemble import VotingRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

voting_model_optimized = VotingRegressor(
    estimators=[
        ('rf', best_rf),
        ('lr', lr_model)
    ]
)

voting_model_optimized.fit(X_train, y_train)
y_pred_voting_opt = voting_model_optimized.predict(X_test)

print("Optimize edilmiş Voting Regressor Performansı:")
print(f"MSE: {mean_squared_error(y_test, y_pred_voting_opt):.2f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred_voting_opt):.2f}")
print(f"R2: {r2_score(y_test, y_pred_voting_opt):.4f}")



Optimize edilmiş Voting Regressor Performansı:
MSE: 1257081815.37
MAE: 16750.57
R2: 0.8361


MSE ve MAE değerleri Random Forest tek başına performansından biraz daha kötü görünüyor (Random Forest’ta MSE ~847M idi), ancak Voting Regressor hâlâ iyi sonuç veriyor.

R² biraz azalmış (Random Forest’ta 0.8895’ti), yani açıklanan varyans biraz düşmüş.

Burada Linear Regression modeli Voting’de etkili olmuş olabilir, çünkü Linear Regression genelde daha basit ve bazen tahminlerde sapmaya sebep olabilir.

Yine de toplu modeli stabil ve makul performans sergiliyor.

### Genel Yorum
##### Random Forest Regressor, en iyi performansı gösterdi (R² ~0.89), yani bu veri seti için güçlü ve güvenilir bir model olarak öne çıktı. Hem karmaşık ilişkileri yakalayabiliyor hem de overfitting riski görece düşük.

##### Lineer Regresyon modeli basit yapısı nedeniyle hız ve yorumlanabilirlik avantajı sağladı, ancak karmaşık ilişkileri yakalamada sınırlı kaldı (R² ~0.64). Yine de temel bir referans noktası olarak değerlendirilebilir.

##### KNN Regressor ve Decision Tree Regressor makul performans sergiledi, ancak RF kadar güçlü değiller. KNN, büyük veri ve çok değişken olduğunda yavaş kalabilir.

##### SVR bu veri setinde ve parametrelerle optimize edilmediğinde performansı zayıftı. Optimize edilmesi zaman alıyor ve pratik değil. Özellikle çok sayıda özellik varsa SVR’nin çalışması zorlaşıyor.

##### Voting Regressor, farklı modellerin bir araya gelmesiyle dengeli bir tahmin sağladı. Ancak tek başına RF’nin performansına ulaşamadı.


# 2.BÖLÜM


### 1.GBM algoritmasının mantığı, zayıf öğrenici kavramı
### 2.XGBoost'un farklılıkları (regularizasyon, erken durdurma, paralelleşme)
### 3.Uygulama: XGBRegressor veya XGBClassifier
### 4.Diğer modellerle karşılaştırma

In [59]:
# Hedef değişeken sayısal olduğu için XDBRegressor kullanacağız.
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np


In [61]:
xgb_model = XGBRegressor(random_state=42)
xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)


In [62]:
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print("XGBRegressor Performansı:")
print("MSE:", mse_xgb)
print("MAE:", mae_xgb)
print("R2:", r2_xgb)


XGBRegressor Performansı:
MSE: 802277896.0207651
MAE: 17375.963358839897
R2: 0.8954049944877625


R² skoru oldukça yüksek (≈ %89.5), yani model hedef değişkendeki varyansın büyük bir kısmını açıklayabiliyor.

MAE ve MSE değerleri, diğer regresyon modelleriyle karşılaştırıldığında oldukça düşük çıktı, bu da tahmin hatalarının az olduğunu gösteriyor.

Model, Random Forest ile benzer doğrulukta, ancak daha hızlı çalışıyor ve daha esnek optimize edilebiliyor.



In [66]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np


In [68]:
# GBM modelini oluştur
gbm = GradientBoostingRegressor(random_state=42)
gbm.fit(X_train, y_train)

# Tahmin yap
y_pred_gbm = gbm.predict(X_test)

# Performans metrikleri
mse_gbm = mean_squared_error(y_test, y_pred_gbm)
mae_gbm = mean_absolute_error(y_test, y_pred_gbm)
r2_gbm = r2_score(y_test, y_pred_gbm)

print("GBM Performansı:")
print("MSE:", mse_gbm)
print("MAE:", mae_gbm)
print("R2:", r2_gbm)


GBM Performansı:
MSE: 814410853.8020935
MAE: 17683.204224348596
R2: 0.8938231788578969



XGBoost biraz daha iyi sonuç verdi, özellikle MSE ve R2 skorlarında (daha düşük MSE, daha yüksek R2).

GBM ise biraz daha yavaş ama performansı oldukça yakın.

İki model de güçlü zayıf öğrenici topluluk yöntemleri, fakat XGBoost'ta regularizasyon, erken durdurma ve paralelleşme gibi ek avantajlar bulunuyor, bu yüzden genellikle daha etkili ve hızlı çalışıyor.

Her iki model de House Prices veri setinde başarılı tahminler yapıyor ve diğer klasik modellere göre (Lineer Regresyon, Random Forest vb.) daha iyi sonuçlar veriyor.

In [71]:
from sklearn.model_selection import RandomizedSearchCV
from xgboost import XGBRegressor
import numpy as np

xgb = XGBRegressor(random_state=42, objective='reg:squarederror')

param_dist = {
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7, 10],
    'learning_rate': [0.01, 0.05, 0.1, 0.2],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0],
    'gamma': [0, 0.1, 0.3],
    'reg_alpha': [0, 0.01, 0.1],
    'reg_lambda': [1, 1.5, 2],
}

random_search = RandomizedSearchCV(
    xgb, param_distributions=param_dist, n_iter=30,
    scoring='neg_mean_squared_error', cv=3, verbose=2, random_state=42, n_jobs=-1
)

random_search.fit(X_train, y_train)

print("En iyi parametreler (XGBoost):", random_search.best_params_)

best_xgb = random_search.best_estimator_

y_pred = best_xgb.predict(X_test)

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

print(f"MSE: {mean_squared_error(y_test, y_pred)}")
print(f"MAE: {mean_absolute_error(y_test, y_pred)}")
print(f"R2: {r2_score(y_test, y_pred)}")


Fitting 3 folds for each of 30 candidates, totalling 90 fits
En iyi parametreler (XGBoost): {'subsample': 0.8, 'reg_lambda': 2, 'reg_alpha': 0, 'n_estimators': 300, 'max_depth': 5, 'learning_rate': 0.1, 'gamma': 0.3, 'colsample_bytree': 0.8}
MSE: 666938040.8612074
MAE: 16123.388163527397
R2: 0.913049578666687


Optimize edilmiş XGBoost modelin performansı gayet başarılı görünüyor:

MSE: 666,938,040 — Öncekinden daha düşük, yani hata azalmış.

MAE: 16,123 — Ortalama hata da azalmış.

R2: 0.913 — Model veriyi çok iyi açıklıyor, %91.3 varyansı yakalamış.

Bu sonuç, önceki modellerle karşılaştırıldığında oldukça iyi bir iyileşme sağladı. XGBoost’un regularizasyon, erken durdurma ve güçlü ağaç tabanlı yapı özellikleri burada faydasını gösteriyor.

### Genel Yorum:
Lineer Regresyon:
R2 ~0.64 — Temel doğrusal model, makul ama sınırlı başarı.

KNN Regressor:
R2 ~0.70 — Komşuluk temelli biraz daha iyi ama yine sınırlı.

Decision Tree Regressor:
R2 ~0.76 — Tek ağaç ile daha iyi sonuç.

Random Forest Regressor (optimize edilmiş):
R2 ~0.89 — Çok sayıda ağacın birleşimiyle yüksek başarı.

SVR:
R2 negatif çıktı, yani kötü performans.

Voting Regressor:
R2 ~0.74 — Model kombinasyonu, orta seviyede.

GBM:
R2 ~0.89 — Random Forest’a yakın, güçlü ağaç tabanlı model.

XGBoost (optimize edilmiş):
R2 ~0.91 — En iyi performans, hatalar en düşük seviyede.

Genel değerlendirme:
XGBoost, hataları en iyi minimize eden ve veri varyansını en iyi açıklayan model olarak öne çıkıyor. Ardından Random Forest ve GBM geliyor. Doğrusal modeller ve KNN bu problem için yeterince esnek değil. SVR ise başarısız oldu.

Bu sonuçlar, ağaç tabanlı yöntemlerin karmaşık ve yüksek boyutlu House Prices veri setlerinde daha başarılı olduğunu gösteriyor. Ayrıca XGBoost’un optimize edilmesiyle performans artışı net bir şekilde ortaya çıktı.

### 1. GBM Algoritmasının Mantığı ve Zayıf Öğrenici Kavramı
Gradient Boosting Machine (GBM), topluluk öğrenmesi (ensemble learning) yöntemlerinden biridir ve zayıf öğrenicilerin (weak learners) ardışık olarak eğitilmesine dayanır. Burada zayıf öğrenici, genellikle tek başına çok güçlü olmayan ama biraz daha iyi tahmin yapabilen basit modellerdir; örneğin küçük derinlikli karar ağaçları.

GBM, ilk olarak basit bir model (zayıf öğrenici) eğitir ve bu modelin hatalarını (residuals) hesaplar.

Sonra, ikinci model bu hataları öğrenmeye çalışır; yani ilk modelin yapamadığını düzeltecek şekilde eğitilir.

Bu süreç ardışık olarak devam eder, her yeni model öncekinin hatalarını azaltmak için eklenir.

Sonuçta, zayıf öğrenicilerin birleşimiyle güçlü ve daha doğru bir model ortaya çıkar.

Bu yöntem, özellikle karmaşık ve doğrusal olmayan ilişkilere sahip verilerde başarılıdır.

### 2. XGBoost’un GBM’den Farkları
XGBoost, GBM algoritmasının geliştirilmiş ve optimize edilmiş halidir. Temel farkları şunlardır:

Regularizasyon (L1 ve L2): XGBoost, model karmaşıklığını kontrol etmek için regularizasyon terimleri ekler. Böylece aşırı uyum (overfitting) azaltılır.

Erken Durdurma (Early Stopping): Model, doğrulama setindeki performans düşmeye başladığında eğitim otomatik olarak durdurulabilir. Bu da aşırı uyumu önler ve eğitim süresini kısaltır.

Paralel İşlem: XGBoost, ağaç oluşturma işlemini paralel yaparak çok daha hızlı çalışır.

Eksik Veri Desteği: XGBoost, eksik değerlerle doğal olarak başa çıkabilir, onları özel olarak işler.

Daha İyi Optimizasyon: İkinci türev tabanlı optimizasyon yaparak daha hızlı ve etkili öğrenme sağlar.

Sütun Örneklemesi: Rastgele olarak ağaç oluştururken kullanılan özelliklerin bir kısmını seçerek çeşitliliği artırır ve aşırı uyumu azaltır.


